pandas/core/frame.py

```diff
@@ -2990,15 +2990,24 @@ def duplicated(self, subset=None, keep='first'):
         from pandas.core.groupby import get_group_index
         from pandas.core.algorithms import factorize
         from pandas.hashtable import duplicated_int64, _SIZE_HINT_LIMIT
+        from pandas.compat import long
 
         def f(vals):
-
-            # if we have integers we can directly index with these
+            # if we have integers we may be able to directly index with these
             if com.is_integer_dtype(vals):
-                from pandas.core.nanops import unique1d
-                labels, shape = vals, unique1d(vals)
-            else:
-                labels, shape = factorize(vals, size_hint=min(len(self), _SIZE_HINT_LIMIT))
+                vals = vals.astype('i8', copy=False)
```

Comment:
I'm trying to avoid undoing #10917, because factorizing the integers is slower than using them directly. If all of the integers are non-negative, then everything actually works perfectly just replacing shape with max + 1. If some of the numbers are negative, then we'd like to just shift them up to all be non-negative, but that might cause an overflow that we need to check for.